### ISSUE TITLE
Crash when saving backup leaves app unresponsive

### ISSUE THREAD

**Issue by charignon** [HappenApps/Quiver#461]
Saving a backup to iCloud drive crashes with the following stacktrace and the app does not recover, I had to kill -9 it.

-[NSVBSavePanel observeValueForKeyPath:ofObject:change:context:] caught non-fatal NSInternalInconsistencyException 'unexpected class type for sandbox extension string!' with backtrace (
        0   CoreFoundation                      0x00007fff95059bd2 **exceptionPreprocess + 178
        1   libobjc.A.dylib                     0x00007fff960e6dd4 objc_exception_throw + 48
        2   CoreFoundation                      0x00007fff9505e8aa +[NSException raise:format:arguments:] + 106
        3   Foundation                          0x00007fff9d96ff46 -[NSAssertionHandler handleFailureInMethod:object:file:lineNumber:description:] + 198
        4   AppKit                              0x00007fff973710bc -[NSVBSavePanel _attachSandboxExtensions:toURL:orURLs:] + 354
        5   AppKit                              0x00007fff9737146d -[NSVBSavePanel _URLWithSecurityScoped:] + 232
        6   AppKit                              0x00007fff96d3c2f8 -[NSVBSavePanel completeWithReturnCode:url:urls:] + 460
        7   AppKit                              0x00007fff96cf935d -[NSVBSavePanel observeValueForKeyPath:ofObject:change:context:] + 594
        8   Foundation                          0x00007fff9d8c51f7 NSKeyValueNotifyObserver + 379
        9   Foundation                          0x00007fff9d8c4a67 NSKeyValueDidChange + 457
        10  Foundation                          0x00007fff9d89139c -[NSObject(NSKeyValueObserverNotification) didChangeValueForKey:] + 118
        11  ViewBridge                          0x00007fff9e3523db withHintInProgress + 501
        12  ViewBridge                          0x00007fff9e2ed695 -[NSViewBridge setObject:forKey:withKVO:] + 899
        13  ViewBridge                          0x00007fff9e2ec1d0 -[NSViewBridge nonLocalChangeInProgress:block:] + 417
        14  ViewBridge                          0x00007fff9e340943 -[NSRemoteViewMarshal exceptionSafeSetRemoteObject:forKey:withReply:] + 275
        15  ViewBridge                          0x00007fff9e340ad1 -[NSRemoteViewMarshal setRemoteObject:forKey:withReply:] + 56
        16  CoreFoundation                      0x00007fff94fcb8cc __invoking_** + 140
        17  CoreFoundation                      0x00007fff94fcb75e -[NSInvocation invoke] + 286
        18  ViewBridge                          0x00007fff9e3523db withHintInProgress + 501
        19  ViewBridge                          0x00007fff9e352161 **deferNSXPCInvocationOntoMainThread_block_invoke_2 + 220
        20  ViewBridge                          0x00007fff9e2f0c3f +[NSViewServiceApplication withHostPID:invoke:] + 46
        21  ViewBridge                          0x00007fff9e35202f __deferNSXPCInvocationOntoMainThread_block_invoke + 129
        22  ViewBridge                          0x00007fff9e351c3a __deferBlockOntoMainThread_block_invoke_2 + 448
        23  CoreFoundation                      0x00007fff94fef24c __CFRUNLOOP_IS_CALLING_OUT_TO_A_BLOCK** + 12
        24  CoreFoundation                      0x00007fff94fceb65 __CFRunLoopDoBlocks + 341
        25  CoreFoundation                      0x00007fff94fce8cb __CFRunLoopRun + 2347
        26  CoreFoundation                      0x00007fff94fcdd38 CFRunLoopRunSpecific + 296
        27  HIToolbox                           0x00007fff9a8ccd55 RunCurrentEventLoopInMode + 235
        28  HIToolbox                           0x00007fff9a8ccb8f ReceiveNextEventCommon + 432
        29  HIToolbox                           0x00007fff9a8cc9cf _BlockUntilNextEventMatchingListInModeWithFilter + 71
        30  AppKit                              0x00007fff96a8df3a _DPSNextEvent + 1067
        31  AppKit                              0x00007fff96a8d369 -[NSApplication _nextEventMatchingEventMask:untilDate:inMode:dequeue:] + 454
        32  AppKit                              0x00007fff96a81ecc -[NSApplication run] + 682
        33  AppKit                              0x00007fff96a4b162 NSApplicationMain + 1176
        34  Quiver                              0x0000000103712934 Quiver + 6452
        35  ???                                 0x0000000000000001 0x0 + 1
    )

**Comment by ylian** [comment]
It's probably just taking a while copying the library. How big is the library?

**Comment by charignon** [comment]
Nope, I waited 10 minutes and it had one note, no attachment. I tried to repro and it copied it instantaneously.
You should probably add a timeout or at least some kind of check that progress is happening.
What's annoying in that case is that it crashed the UI thread.

**Comment by ylian** [comment]
Yep it should not block the UI thread. I will fix.

**Comment by Iomegan** [comment]
I have the same issue with one of my products as well. I figured out that it always happens if I set the DirectoryURL property of the NSSavePanel instance. However, some users report that it still crashes (or hangs up) without having the DirectoryURL property set. Also I wasn't able to reproduce it in a test project. Very strange...

**Comment by Iomegan** [comment]
Okay, I think I could fix it. The problem for me was that  i was setting a file path as the directoryURL. i.e. ~/Documents/test.txt. Then I removed the line that set the (wrong) directoryURL and let some customers test the new build. However if you don't set the directoryURL in code, it will take the string from NSUSerDefaults's standardDefaults _NSNavLastRootDirectory_. And that was still the old wrong file path. So the solution is simply to make sure you set a directory not a file path. In my case I remove the last PathComponent from the URL I use to set the SavePanel's directoryURL property.
`[savePanel setDirectoryURL:fileURL.URLByDeletingLastPathComponent];`

**Comment by nicolas-miari** [comment]
@Iomegan You saved my life. I have nothing to do with this project (Quiver), but a search of "_attachSandboxExtensions:toURL:orURLs" led me here: I was doing the exact same thing, for similar reasons (i.e., export a modified file to the same directory it was originally read from). However, I only experienced the crash when enabling sandbox in my app (which I need, to publish to the AppStore).

**Comment by shieladixon** [comment]
I also had users experiencing a crash in an unrelated app. (with _attachSandboxExtensions:toURL:orURLs:) I'm incredibly grateful to Iomegan for the answer here. In my testing, an item with key NSNavLastRootDirectory did exist in the standardUserDefaults, but it was already a directory (eg ~/Downloads or wherever the last save happened). My app was creating a new file, not one that had been opened.   Taking NSNavLastRootDirectory from defaults (if it exists - or if not, using a suitable default like ~/Downloads), making it into a NSURL with URLWithString: and sending that to  savePanel setDirectoryURL:  after creating the save panel seems to have cured my crash. I can't help feeling that this is a bug in the system.

**Comment by michaeladamkatz** [comment]
See also https://github.com/mlabbe/nativefiledialog/issues/76 for the same crash, but caused by not having called setFieldStringValue. Same situation: only in MacOS 10.15, only when trying to save a file twice to the same directory, and only when using the default savePanel settings (not changing directory, file name, etc.).

**Comment by IhorKaravan** [comment]
I have tried to extend NSOpenSavePanelDelegate to create my own delegate. 
@interface CocoaSaveDlgDelegate : NSObject <NSApplicationDelegate, NSOpenSavePanelDelegate>
@end

Then set this delegate for savePanel: [savePanel setDelegate: [CocoaSaveDlgDelegate alloc]]. In this delegate, I need to override validateURL and return always YES.

@implementation CocoaSaveDlgDelegate

- (BOOL)panel:(id)sender validateURL:(NSURL *)url error:(NSError **)outError 
{
    return YES;
}

@end

This solution allows re-write existing file without a crash, but I don't know if is secure. It can be a bad idea to do it.
